flatpak: Allow generating the Flatpak bundle

livingsilver94 opened the pull request:
#### What it does
This PR allows to generate a Flatpak bundle of theia-ide.

This PR does *not* publish Theia on Flathub. Publishing applications to Flathub requires a process in which building the Flatpak bundle is only one of the steps. I'll try to take care of it later.

This PR is related to #360, although it doesn't entirely solve it because Theia is not available on Flathub yet.
Closes #527: Theia escapes the sandbox by default. To prevent that, run it after setting the environment variable `THEIA_NO_ESCAPE_SANDBOX` to any value, e.g. `1`.

#### How to test
You should follow the Readme. After that, to run Theia, either run it from the application menu of your desktop environment or launch `flatpak run org.eclipse.theia`.

At any rate, here is the convenient command to both build and install Theia: `flatpak-builder flatpak/build flatpak/org.eclipse.theia.yaml --install-deps-from=flathub --force-clean --user --install` (note the `--install` flag).

If you run `flatpak run org.eclipse.theia`, Theia will escape the Flatpak sandbox by default. To check that, open the terminal. You should see your usual prompt, environment variables, and you should be able to browse the entire system. If you set `THEIA_NO_ESCAPE_SANDBOX`, the terminal should be limited.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

sgraband commented:
Hey @livingsilver94, thank you for your contribution! Could you please sign the [Eclipse ECA](https://www.eclipse.org/legal/eca/) and sign your commits with the mail you signed the ECA, so we can take a look at the PR?

livingsilver94 commented:
I don't know what's going on here.

- In [this page](https://api.eclipse.org/git/eca/status/gh/eclipse-theia/theia-ide/617), the "ECA Validation Tool" reports I have signed the ECA.
- My commit is signed-off with correct email address.
- I have linked my GH account with the Eclipse Foundation account.

Yet the autocheck says I haven't signed the ECA, even after I pressed "revalidate".

livingsilver94 commented:
It worked now. I did absolutely nothing. It seems it just took some time.

livingsilver94 commented on flatpak/org.eclipse.theia.metainfo.xml:
```diff
@@ -0,0 +1,56 @@
+    <display_length compare="ge">768</display_length>
+  </requires>
+
+  <url type="bugtracker">https://github.com/eclipse-theia/theia-ide/issues</url>
```
Do you prefer theia-ide or theia?

livingsilver94 commented:
I've rebased my PR against master.

LorbusChris commented:
I personally think the flatpak id should be `org.eclipse.theia-ide` for clarity, to avoid confusion with the `theia` repo.

Also, could we default to wayland? Running with it works fine now.

I've opened https://github.com/eclipse-theia/theia-ide/pull/635 as a follow up to escape the sandbox and bundle all default extensions.

LorbusChris commented on flatpak/org.eclipse.theia.yaml:
```diff
@@ -0,0 +1,71 @@
+      - type: script
+        dest-filename: wrapper.sh
+        commands:
+          - exec zypak-wrapper /app/share/theia/theia-ide-electron-app.bin --ozone-platform=x11 --no-sandbox "$@"
```
```suggestion
          - exec zypak-wrapper /app/share/theia/theia-ide-electron-app.bin --ozone-platform-hint=auto --no-sandbox "$@"
```

LorbusChris commented on flatpak/org.eclipse.theia.metainfo.xml:
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+  <id>org.eclipse.theia</id>
```
I think we still want at least a URL redirect to theia-ide.org there (best on both https://theia.eclipse.org and https://theia-ide.eclipse.org)

livingsilver94 replied:
See https://github.com/flathub/flathub/pull/6473#discussion_r2162317717

LorbusChris commented:
Using the wrapper script for escaping as implemented in https://github.com/eclipse-theia/theia-ide/pull/635/changes/ed46c78fd299d742e504f7809a740cf455f51882 seems cleaner to me than doing it from JS.

LorbusChris commented on flatpak/org.eclipse.theia.yaml:
```diff
@@ -0,0 +1,72 @@
+  - --env=XCURSOR_PATH=/run/host/user-share/icons:/run/host/share/icons
+  - --filesystem=host
+  - --persist=.theia-ide
+  - --socket=fallback-x11
```
nit: sockets should be grouped together.

LorbusChris commented on flatpak/theia.sh:
```diff
@@ -0,0 +1,42 @@
+# settings and workspaces won't be shared.
+# Since this is a Flatpak, everything will be stored within the Flatpak directory structure.
+
+default_args="--ozone-platform-hint=auto --electronUserData=$XDG_DATA_HOME"
```
The electronUserData flag does not seem to be necessary when passing the XDG envs as as a whole:
```
--env=XDG_DATA_HOME="${XDG_DATA_HOME:-}" \
--env=XDG_CONFIG_HOME="${XDG_CONFIG_HOME:-}" \
--env=XDG_CACHE_HOME="${XDG_CACHE_HOME:-}" \
 --env=XDG_STATE_HOME="${XDG_STATE_HOME:-}" \
```

Not sure what the preferred way is here.

LorbusChris commented on applications/electron/package.json:
```diff
@@ -5,6 +5,7 @@
   "productName": "Theia IDE",
   "version": "1.67.100",
   "main": "scripts/theia-electron-main.js",
+  "desktopName": "org.eclipse.theia.desktop",
```
ah well, I see you've also hit the issue now. Ref https://github.com/eclipse-theia/theia-ide/pull/635#discussion_r2649106295

LorbusChris replied:
Please double check if that can be added here safely, or whether that breaks assumptions made by AppImage or deb builds. (Alternately this could be patched in the flatpak build itself)

livingsilver94 commented on flatpak/theia.sh:
```diff
@@ -0,0 +1,42 @@
+
+launch_sandboxed() {
+    default_args="$default_args --no-sandbox"
+    THEIA_CONFIG_DIR="$XDG_CONFIG_HOME" \
```
It's required. Grep that env in https://github.com/eclipse-theia/theia.

LorbusChris replied:
I don't think Flathub cares as long as there's an official backing to the Flatpak maintenance. But I do see your point of wanting it to be the same dir for running inside the sandbox and escaped from it., though.

LorbusChris commented:
re https://github.com/eclipse-theia/theia-ide/pull/617#discussion_r2653082922: https://github.com/eclipse-theia/theia-ide/pull/617#discussion_r2653106447

LorbusChris reviewed (approved):
This looks mostly good now.

Eclipse folks will need to clarify the desired Flatpak name, and possibly a URL redirect to theia-ide.org (see https://github.com/eclipse-theia/theia-ide/pull/617/changes#r2649892938). Alternatively a name such as `org.theia-ide.theia` could be used.

LorbusChris commented on flatpak/theia.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/sh
+
```
let's add `set -xeuo pipefail` here per the Unofficial Bash Strict Mode recommendations (http://redsymbol.net/articles/unofficial-bash-strict-mode/)

livingsilver94 commented:
Are you still interested in the Flatpak support, @sgraband? If so, I need to rebase this PR.

sgraband commented:
@ndoschek Could you answer here?

ndoschek commented:
Hi @livingsilver94, thanks for your patience here and sorry for the delayed response!

Yes, we are still interested in Flatpak support for Theia IDE. We've discussed this internally and as a first step I've opened a helpdesk ticket with the Eclipse Foundation to get guidance on the Flathub publishing process from a policy/security perspective. Once we have the green light, we plan to set up proper CI infrastructure and move forward with the Flathub submission.

We really appreciate your interest and work on this. We'll keep you posted as things progress!

fgiancane8 commented:
sorry for jumping in the conversation. as a Fedora Linux user (possibly also the immutable version of the distro) implementing a flatpak deployment for Theia would definitely help adoption! Also supporting the native escape from the sandbox (to use host-native development tools) would be super useful. Thanks for all the effort and I am available to test the changes as a user, in case is needed!

livingsilver94 commented:
I rebased this PR but I couldn't build the Flatpak bundle because of https://github.com/eclipse-theia/theia-ide/issues/742#issuecomment-5140702211.

livingsilver94 committed:
flatpak: Allow generating the Flatpak bundle

This PR allows to generate a Flatpak bundle of theia-ide.

This PR does not publish Theia on Flathub. Publishing applications to Flathub
requires a process in which building the Flatpak bundle is only one of the steps.

Signed-off-by: Fabio Forni <development@redaril.me>

livingsilver94 committed:
flatpak: Escape the Flatpak sandbox when requested

Signed-off-by: Fabio Forni <development@redaril.me>

livingsilver94 committed:
applications: electron: Set desktopName in package.json

This is required in order to show the Theia icon when run outside the Flatpak
sandbox.
This was originally authored by Christian Glombek in cbed2d3bd7c9ead4285660bd20cd7b5e257e575f.

Co-authored-by: Christian Glombek <c.glombek@cosa.systems>
